### DISCUSSION TITLE
Install Fails on Debian 10 / error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

### DISCUSSION THREAD

**Discussion by GSY-PI** [apache/airflow#21035] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

Error during install  compilation terminated. error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

Full trace:

Collecting apache-airflow[airbyte,amazon,async,celery,docker,facebook,ftp,google,google_auth,mongo,mysql,neo4j,oracle,postgres,sendgrid,sftp,slack,ssh]==2.2.3
  Using cached https://files.pythonhosted.org/packages/94/45/5c95e3f83ea09fbc6d655b76254dcfd5d29a18ad5abfa7ef4c2151334c81/apache_airflow-2.2.3-py3-none-any.whl
Collecting cryptography==3.4.8 (from -c https://raw.githubusercontent.com/apache/airflow/constraints-2.2.3/constraints-3.7.txt (line 203))
  Using cached https://files.pythonhosted.org/packages/cc/98/8a258ab4787e6f835d350639792527d2eb7946ff9fc0caca9c3f4cf5dcfe/cryptography-3.4.8.tar.gz
  Installing build dependencies ... error
  Complete output from command /home/revrebel/airflow/airflow-env/bin/python3.7 -m pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-ac9djfhx --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools>=40.6.0 wheel "cffi>=1.12; platform_python_implementation != 'PyPy'" setuptools-rust>=0.11.4:
  Collecting setuptools>=40.6.0
    Using cached https://files.pythonhosted.org/packages/eb/53/0dd4c7960579da8be13fa9b2c2591643d37f323e3d79f8bc8b1b6c8e6217/setuptools-60.5.0-py3-none-any.whl
  Collecting wheel
    Using cached https://files.pythonhosted.org/packages/27/d6/003e593296a85fd6ed616ed962795b2f87709c3eee2bca4f6d0fe55c6d00/wheel-0.37.1-py2.py3-none-any.whl
  Collecting cffi>=1.12
    Using cached https://files.pythonhosted.org/packages/00/9e/92de7e1217ccc3d5f352ba21e52398372525765b2e0c4530e6eb2ba9282a/cffi-1.15.0.tar.gz
  Collecting setuptools-rust>=0.11.4
    Using cached https://files.pythonhosted.org/packages/66/ca/66bdf8f326977098eff28c314c8f825bc28d6986944c590e40ad0f74c5f0/setuptools_rust-1.1.2-py3-none-any.whl
  Collecting pycparser (from cffi>=1.12)
    Using cached https://files.pythonhosted.org/packages/62/d5/5f610ebe421e85889f2e55e33b7f9a6795bd982198517d912eb1c76e1a53/pycparser-2.21-py2.py3-none-any.whl
  Collecting semantic-version<3,>=2.8.2 (from setuptools-rust>=0.11.4)
    Using cached https://files.pythonhosted.org/packages/a5/15/00ef3b7888a10363b7c402350eda3acf395ff05bebae312d1296e528516a/semantic_version-2.8.5-py2.py3-none-any.whl
  Collecting typing-extensions>=3.7.4.3 (from setuptools-rust>=0.11.4)
    Using cached https://files.pythonhosted.org/packages/05/e4/baf0031e39cf545f0c9edd5b1a2ea12609b7fcba2d58e118b11753d68cf0/typing_extensions-4.0.1-py3-none-any.whl
  Building wheels for collected packages: cffi
    Running setup.py bdist_wheel for cffi: started
    Running setup.py bdist_wheel for cffi: finished with status 'error'
    Complete output from command /home/revrebel/airflow/airflow-env/bin/python3.7 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-oq_islxp/cffi/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-01r9ap7u --python-tag cp37:
    usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: -c --help [cmd1 cmd2 ...]
       or: -c --help-commands
       or: -c cmd --help
  
    error: invalid command 'bdist_wheel'
  
    ----------------------------------------
    Failed building wheel for cffi
    Running setup.py clean for cffi
  Failed to build cffi
  Installing collected packages: setuptools, wheel, pycparser, cffi, semantic-version, typing-extensions, setuptools-rust
    Running setup.py install for cffi: started
      Running setup.py install for cffi: finished with status 'error'
      Complete output from command /home/revrebel/airflow/airflow-env/bin/python3.7 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-oq_islxp/cffi/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-yxg8fh5y/install-record.txt --single-version-externally-managed --prefix /tmp/pip-build-env-ac9djfhx --compile --install-headers /home/revrebel/airflow/airflow-env/include/site/python3.7/cffi:
      running install
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.7
      creating build/lib.linux-x86_64-3.7/cffi
      copying cffi/api.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/vengine_gen.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/__init__.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/setuptools_ext.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/backend_ctypes.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/recompiler.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/cparser.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/lock.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/ffiplatform.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/commontypes.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/pkgconfig.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/verifier.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/cffi_opcode.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/error.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/vengine_cpy.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/model.py -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/_cffi_include.h -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/parse_c_type.h -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/_embedding.h -> build/lib.linux-x86_64-3.7/cffi
      copying cffi/_cffi_errors.h -> build/lib.linux-x86_64-3.7/cffi
      running build_ext
      building '_cffi_backend' extension
      creating build/temp.linux-x86_64-3.7
      creating build/temp.linux-x86_64-3.7/c
      x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/ffi -I/usr/include/libffi -I/home/revrebel/airflow/airflow-env/include -I/usr/include/python3.7m -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.7/c/_cffi_backend.o
      c/_cffi_backend.c:15:10: fatal error: ffi.h: No such file or directory
       #include <ffi.h>
                ^~~~~~~
      compilation terminated.
      error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  
      ----------------------------------------
  Command "/home/revrebel/airflow/airflow-env/bin/python3.7 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-oq_islxp/cffi/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-yxg8fh5y/install-record.txt --single-version-externally-managed --prefix /tmp/pip-build-env-ac9djfhx --compile --install-headers /home/revrebel/airflow/airflow-env/include/site/python3.7/cffi" failed with error code 1 in /tmp/pip-install-oq_islxp/cffi/
  
  ----------------------------------------
Command "/home/airflow/airflow-env/bin/python3.7 -m pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-ac9djfhx --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools>=40.6.0 wheel "cffi>=1.12; platform_python_implementation != 'PyPy'" setuptools-rust>=0.11.4" failed with error code 1 in None

### What you expected to happen

To install

### How to reproduce

When using the following:

AIRFLOW_VERSION=2.2.3
PYTHON_VERSION="$(python --version | cut -d " " -f 2 | cut -d "." -f 1-2)"
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-{PYTHON_VERSION}.txt"
INSTALL_PROVIDERS_FROM_SOURCES="true"
pip3 install "apache-airflow[async,celery,postgres,google,amazon,airbyte,facebook,ftp,google_auth,mongo,mysql,neo4j,oracle,sendgrid,sftp,slack,ssh,docker]==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"

### Operating System

PRETTY_NAME="Debian GNU/Linux 10 (buster)" NAME="Debian GNU/Linux" VERSION_ID="10" VERSION="10 (buster)" VERSION_CODENAME=buster ID=debian HOME_URL="https://www.debian.org/" SUPPORT_URL="https://www.debian.org/support" BUG_REPORT_URL="https://bugs.debian.org/"

### Versions of Apache Airflow Providers

Not installed yet

### Deployment

Virtualenv installation

### Deployment details

Linode Virtual Server

### Anything else

No

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Reply by potiuk** [reply]
This is not airlfow error - that's why converted it into discussion. 

You likely miss wheel package that is needed to compile some external packages (Building wheels for collected packages: cffi) -because there are no binary packages distributed by the maintainers of the packages.

In this case the package in question is this:

> Building wheels for collected packages: cffi

This is standard approach with installing python packages that cannot be installed. You should look for solution related to this package - either in the place where support of `cffi` is or just looking in Google.

My best guess is that just `pip install wheel` might help, but there might be other problems - so do not take it for granted.

**Comment by GSY-PI** [comment]
Sorry you close this, fixed by installing libffi-dev which you may want to note on the setup guide.

**Reply by potiuk** [reply]
It really depends on the system you install it on. There is no "generic" rule for such dependncies - depending on distribution the prerequisites might be different. 

But if you feel like it might help - you are free to make Pull Request. It's extremely easy - even easier than creating issue.  Airflow is created by > 1800 contributors so you can become one of them 

For example here: https://airflow.apache.org/docs/apache-airflow/stable/installation/prerequisites.html you just click "Suggest a change on this page" at bottom left and PR will be opened where you can propose your changes. As a person who struggled with it, you are likely the best one to describe it in the way that will be helpful for others. 

I really encourage you to do that!
